File: src/libs/m3/netrs/NetChannel.cc
/*
 * Copyright (C) 2021, Tendsin Mende <tendsin.mende@mailbox.tu-dresden.de>
 * Copyright (C) 2017, Georg Kotheimer <georg.kotheimer@mailbox.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <base/log/Lib.h>

#include <m3/netrs/Net.h>
#include <m3/netrs/NetChannel.h>

namespace m3 {

NetChannel::NetChannel(capsel_t caps)
    : _sg(SendGate::bind(caps + 1, nullptr)),
      _rg(RecvGate::bind(caps + 0, nextlog2<MSG_BUF_SIZE>::val, nextlog2<MSG_SIZE>::val)),
      _mem(MemGate::bind(caps + 2)) {
    // Activate the rgate manually
    _rg.activate();
}

void NetChannel::send(m3::net::NetData data) {
    LLOG(NET, "NetLogSend:");
    data.log();
    _sg.send_aligned(&data, data.send_size());
}

m3::net::NetData *NetChannel::receive() {
    const TCU::Message *msg = _rg.fetch();
    if(msg != nullptr) {
        LLOG(NET, "msglength=" << msg->length << " sizeof=" << sizeof(m3::net::NetData));
        // this is an actual package, therefore copy the data into a buffer thats cast
        // into the NetData struct
        m3::net::NetData *package = new m3::net::NetData();
        // TODO Somehow prevent copy?
        memcpy(static_cast<void *>(package), msg->data, msg->length);
        // package->log();
        // Ack message to free channel
        _rg.ack_msg(msg);
        return package;
    }
    else {
        return nullptr;
    }
}

}

File: src/libs/rust/m3/src/net/net_channel.rs
/*
 * Copyright (C) 2021, Tendsin Mende <tendsin.mende@mailbox.tu-dresden.de>
 * Copyright (C) 2017, Georg Kotheimer <georg.kotheimer@mailbox.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

use crate::cap::Selector;
use crate::com::{MemGate, RecvGate, SendGate};
use crate::errors::{Code, Error};
use crate::net::NetData;

use super::{MSG_BUF_ORDER, MSG_ORDER};

pub struct NetChannel {
    sg: SendGate,
    rg: RecvGate,

    #[allow(dead_code)]
    mem: MemGate, // TODO Used when socket as file is used?
}

impl NetChannel {
    /// Creates a new channel that is bound to `caps` and `caps+2`. Assumes that the `caps` where obtained from the netrs service, and are valid gates
    pub fn new_with_gates(send: SendGate, mut recv: RecvGate, mem: MemGate) -> Self {
        // activate rgate
        recv.activate().expect("Failed to activate server rgate");

        NetChannel {
            sg: send,
            rg: recv,
            mem,
        }
    }

    /// Does not crate new gates for this channel, but binds to them at `caps`-`caps+2`
    pub fn bind(caps: Selector) -> Result<Self, Error> {
        let mut rgate = RecvGate::new_bind(caps + 0, MSG_BUF_ORDER, MSG_ORDER);
        rgate.activate().expect("Failed to activate rgate");
        let sgate = SendGate::new_bind(caps + 1);
        let mgate = MemGate::new_bind(caps + 2);

        Ok(NetChannel {
            sg: sgate,
            rg: rgate,
            mem: mgate,
        })
    }

    /// Sends data over the send gate
    pub fn send(&self, net_data: NetData) -> Result<(), Error> {
        self.sg.send_aligned(
            &net_data as *const _ as *const u8,
            net_data.send_size(),
            &self.rg,
        )?;
        Ok(())
    }

    /// Tries to receive a message from the other side
    pub fn receive(&self) -> Result<NetData, Error> {
        // Fetch message by hand, if something is fetched,
        // assumes that it is a NetData package.
        if let Some(msg) = self.rg.fetch() {
            // TODO can we get around the clone?
            // safety: we know that we always receive NetData here and when receiving it, it doesn't
            // have to be 2048-byte aligned.
            let net_data = unsafe { msg.get_data_unchecked::<NetData>().clone() };
            // mark message as read
            self.rg.ack_msg(msg)?;

            Ok(net_data)
        }
        else {
            Err(Error::new(Code::NotSup))
        }
    }
}
